### PR TITLE
Migrate FileStorage from Java to Kotlin

### PR DESCRIPTION
**Pull request by MiSikora**
## Description

This is a part of Java to Kotlin migration outlined in #44. This PR migrates [FileStorage.java](https://github.com/Automattic/pocket-casts-android/blob/61fe9bf02411dbb158ec838d2fc4834099e99a4b/modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.java). I kept logic and mostly didn't "Kotlinify" it here. I have a separate [branch](https://github.com/Automattic/pocket-casts-android/tree/task/file-storage-kotlin-improvements) with improvements. I'll make a separate PR after this one gets positively reviewed.

The main changes that are not related directly to migration are dead code removals and some minor fixes between 18b35c1f26b9e55f66a3f604d3323bb5e923b3be...3bdf60042ab98271db8c822dc467699342c6df7d.

The migration process is between 1bb2bfd3fd8e649ad58d08ad98ea3ef3de4b3619...1b0deab348d00842408222652eed5ca28bae0399. For the ease of review I recommend reviewing commits between separately. Each method is migrated separately so the overload should be minimal.

fe739091e6ea4c24c57c12dc75e4019cef733052 is a "dummy" commit to keep track of old changes to the Java file. Otherwise Kotlin migration would delete the history as git doesn't recognize both moving a file `.java -> .kt` and changing file content as the same file.

Commits that follow after fe739091e6ea4c24c57c12dc75e4019cef733052 are simple changes either due to compiler requirements after migration or some nitpicky stuff from my side where I wanted to unify naming.

## Testing Instructions

Unfortunately I'm not really sure how to properly test it.

My first test was with a fresh app if it works and so on. I was able to download the episodes and listen to them offline, which is fine smoke test. However, I also tested it by having an old app and checking if the changes do not mess anything. I was still able to listen to downloaded episodes and so on.

However, there is also an issue of migrating storage and Android 10+ devices. I tested it with emulator since I don't have a physical device.

## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [ ] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...

- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

### CODE REVIEW

**Comment by pocketcasts** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by pocketcasts** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+                    val dotPosition = fileName.lastIndexOf('.')
+                    if (dotPosition < 1) {
+                        return@fileIteration
+                    }
+                    val uuid = fileName.substring(0, dotPosition)
+                    if (uuid.length != 36) {
+                        return@fileIteration
+                    }
+
+                    @Suppress("DEPRECATION")
```
Comment (Mukuljangir372) [inline modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt:266]: Should we remove deprecated code?

Comment (MiSikora) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt:266]: Yes, but I didn't want to introduce such changes in this PR to reduce the review overhead only to pure code migration. Maybe with some minor improvements. Removing this deprecation requires introducing a suspending modifier to the function. I have a separate [branch](https://github.com/Automattic/pocket-casts-android/tree/task/file-storage-kotlin-improvements) with such improvements.

Comment (mchowning) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt:266]: > I didn't want to introduce such changes in this PR to reduce the review overhead only to pure code migration.

Thank you for this! 🙇 🙂

**Review by mchowning — APPROVED** [review]
Thanks for this @MiSikora ! This looks so much better in Kotlin! 😄 

Code changes look good. I left a couple of comments, but they're only minor issues and none of the changes I suggest are blockers, so I'll go ahead and approve.

I tested this on a physical device running Android 13 as well as emulators running Android 7 and Android 9. I downloaded files, went into airplane mode and then moved the files around (to/from sd card and to/from custom folder locations), and I did not observe any playback issues other than [this one small issue](https://github.com/Automattic/pocket-casts-android/issues/1679) that I just logged, but that issue existed before the changes in this PR. I also inspected the device's file system to verify that the files were getting moved around as expected and everything looks good.

Great work! 👍

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+                    LogBuffer.i(LogBuffer.TAG_BACKGROUND_TASKS, "Found downloaded episode ${episode.title}")
+                    val downloadedFilePath = episode.downloadedFilePath?.takeIf(String::isNotBlank)
+                    if (downloadedFilePath == null) {
+                        LogBuffer.e(LogBuffer.TAG_BACKGROUND_TASKS, "Episode had not file path")
+                        return@forEach
+                    }
+                    val file = File(downloadedFilePath)
+                    if (file.exists() && file.isFile) {
+                        moveFileToDir(downloadedFilePath, episodesDir)?.let { updatedPath ->
+                            episodesManager.updateDownloadFilePath(episode, updatedPath, markAsDownloaded = false)
```
[the remark this answers is not in the document]
Comment (MiSikora) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt:205]: Yes, this is correct.

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        val storageCustomDir = File(path)
+        if (!storageCustomDir.exists() && !storageCustomDir.mkdirs()) {
+            throw StorageException("Storage custom folder unavailable.")
+        }
+        storageCustomDir
+    } else {
+        File(choice)
+    }
+
+    private fun createDir(dir: File): File = dir.also(File::mkdirs)
```
[the remark this answers is not in the document]
Comment (MiSikora) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt:91]: Yes, I agree. I have this change on a separate branch [here](https://github.com/Automattic/pocket-casts-android/commit/ee657f3d5fe0dca60e5d9817017f15a80cedc52c). If it is ok I'd skip it here and move it to a subsequent PR?

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+                newDir.mkdirs()
+                val newPocketCastsDir = File(newDir, "PocketCasts")
+                val episodesDir = getOrCreateDir(newPocketCastsDir, DIR_EPISODES)
+
+                // Check existing media and mark those episodes as downloaded
+                if (episodesDir.exists()) {
+                    episodesDir.listFiles()?.forEach { file ->
+                        val fileName = FileUtil.getFileNameWithoutExtension(file)
+                        if (fileName.length < 36) {
```
[the remark this answers is not in the document]
Comment (MiSikora) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt:175]: Similarly to the change with `createDirs()` I have a change related to this on a separate branch [here](https://github.com/Automattic/pocket-casts-android/commit/1559303d4e806144d2f11c37e547fd6952471e1a).

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (mchowning) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt:91]: >  If it is ok I'd skip it here and move it to a subsequent PR?

That's definitely fine! 👍

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (mchowning) [reply modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt:175]: 👍 Feel free to go ahead and merge this.

**Comment by pocketcasts** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by pocketcasts** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**